libtenzir/include/tenzir/detail/default_formatter.hpp

```diff
@@ -0,0 +1,155 @@
+/// format strings `{}` (for user-friendly stringification), `{:?}` (for concise
+/// debug information), and `{:#?}` (for multi-line debug information).
+///
+/// To enable the default formatter for a type `T`, supply a specialization of
+/// `enable_default_formatter` with a `::value` member that is `true`.
+template <class T>
+struct enable_default_formatter : std::false_type {};
+
+namespace detail {
+
+template <class T>
+auto stringify(fmt::format_context::iterator out, const T& x) {
+  if constexpr (printable<decltype(out), T>) {
+    return print(out, x);
+  } else if constexpr (requires { x.to_string(); }) {
+    auto str = x.to_string();
+    return std::copy(str.begin(), str.end(), out);
+  } else if constexpr (requires { to_string(x); }) {
+    auto str = to_string(x);
+    return std::copy(str.begin(), str.end(), out);
+  } else {
+    return;
+  }
+}
+
+template <class T>
+constexpr bool can_stringify
+  = not std::is_same_v<decltype(tenzir::detail::stringify(
+                         std::declval<fmt::format_context::iterator>(),
+                         std::declval<T>())),
+                       void>;
+
+template <class T>
+constexpr bool can_inspect = not std::is_same_v<
+  decltype(caf::inspect_access_type<tenzir::debug_writer, T>()),
+  caf::inspector_access_type::none>;
+
+} // namespace detail
+
+/// Utility class to use the default formatter for arbitrary types.
+template <class T>
+class use_default_formatter {
+public:
+  explicit use_default_formatter(const T& inner) : inner_{inner} {
+  }
+
+  auto to_string() const
+    requires detail::can_stringify<T> || fmt::is_formattable<T>::value
+  {
+    if constexpr (detail::can_stringify<T>) {
+      return detail::stringify(inner_);
+    } else {
+      return fmt::to_string(inner_);
+    }
+  }
+
+  friend auto inspect(auto& f, use_default_formatter& x) -> bool
+    requires detail::can_inspect<T>
+  {
+    return f.apply(x.inner_);
+  }
+
+private:
+  const T& inner_;
+};
+
+template <class T>
+struct enable_default_formatter<use_default_formatter<T>> : std::true_type {};
+
+} // namespace tenzir
+
+template <class T>
+  requires(tenzir::enable_default_formatter<T>::value)
+struct fmt::formatter<T> {
+  enum { normal, debug, pretty_debug } mode = normal;
+
+  static constexpr bool can_stringify = tenzir::detail::can_stringify<T>;
+
+  static constexpr bool can_inspect = tenzir::detail::can_inspect<T>;
+
+  constexpr auto parse(fmt::format_parse_context& ctx)
+    -> fmt::format_parse_context::iterator {
+    auto it = ctx.begin();
+    if (it != ctx.end()) {
+      if (*it == '?') {
+        mode = debug;
+        return ++it;
+      } else if (*it == '#') {
+        auto next = std::next(it);
+        if (next != ctx.end() && *next == '?') {
+          mode = pretty_debug;
+          return ++next;
+        }
+      }
+    }
+    // TODO: Do we want make `{}` use `inspect` if there is no `to_string`?
```

Comment:
This question basically boils down to whether we want to print the debug representation when using `{}` with an object that does not have `to_string()`, or whether we want to reject this at compile-time. The advantage of rejecting it is that we do not return debug representations on accident.